web/app/widgets/timeline/timeline.widget.js

```diff
@@ -0,0 +1,281 @@
+                if (raw.data[i].state !== currentState) {
+                    partitions.push({
+                        state: currentState,
+                        starting_time: currentStartTime,
+                        ending_time: raw.data[i].time
+                    });
+                    currentState = raw.data[i].state;
+                    currentStartTime = raw.data[i].time;
+                }
+            }
+
+            // push the last partition
+            partitions.push({
+                state: currentState,
+                starting_time: currentStartTime,
+                ending_time: raw.data[raw.data.length-1].time
+            });
+
+            return partitions;
+        }
+
+        var startTime = function () {
+            var startDate = new Date();
+            switch (vm.widget.period)
+            {
+                case 'h': startDate.setTime(startDate.getTime() - 60*60*1000); break;
+                case '4h': startDate.setTime(startDate.getTime() - 4*60*60*1000); break;
+                case '8h': startDate.setTime(startDate.getTime() - 8*60*60*1000); break;
+                case '12h': startDate.setTime(startDate.getTime() - 12*60*60*1000); break;
+                case 'D': startDate.setTime(startDate.getTime() - 24*60*60*1000); break;
+                case '3D': startDate.setTime(startDate.getTime() - 3*24*60*60*1000); break;
+                case 'W': startDate.setTime(startDate.getTime() - 7*24*60*60*1000); break;
+                case '2W': startDate.setTime(startDate.getTime() - 2*7*24*60*60*1000); break;
+                case 'M': startDate.setTime(startDate.getTime() - 31*24*60*60*1000); break; //Well...
+                case '2M': startDate.setTime(startDate.getTime() - 2*31*24*60*60*1000); break;
+                case '4M': startDate.setTime(startDate.getTime() - 4*31*24*60*60*1000); break;
+                case 'Y': startDate.setTime(startDate.getTime() - 12*31*24*60*60*1000); break;
+                default: startDate.setTime(startDate.getTime() - 24*60*60*1000); break;
+            }
+            return startDate;
+        }
+        var startDate = startTime();
+
+        if (!vm.widget.series || !vm.widget.series.length)
+            return;
+        
+        vm.rawdata = [];
+        for (var i = 0; i < vm.widget.series.length; i++) {
+            vm.rawdata[i] = $http.get('/rest/persistence/items/' + vm.widget.series[i].item + '?serviceId=' + vm.widget.service + "&boundaries=true&starttime=" + startDate.toISOString());
+        }
+
+        $scope.colorScale = { states: [], colors: [] };
+        for (var i = 0; i < vm.widget.colorMaps.length; i++) {
+            $scope.colorScale.states.push(vm.widget.colorMaps[i].state);
+            $scope.colorScale.colors.push(vm.widget.colorMaps[i].color);
+        }
+
+        $q.all(vm.rawdata).then(function (values) {
+            var partitioned = {};
+            var data = [];
+
+            for (var i = 0; i < values.length; i++) {
```

Comment:
makes no real difference - https://coderwall.com/p/kvzbpa/don-t-use-array-foreach-use-for-instead though! (it might have an impact not in this particular loop but the ones above)